File: api/rxjava/src/main/java/com/coditory/sherlock/rxjava/SherlockWithConnector.java
package com.coditory.sherlock.rxjava;

import com.coditory.sherlock.DistributedLockBuilder;
import com.coditory.sherlock.DistributedLockBuilder.LockCreator;
import com.coditory.sherlock.OwnerIdPolicy;
import com.coditory.sherlock.connector.InitializationResult;
import com.coditory.sherlock.connector.ReleaseResult;
import com.coditory.sherlock.rxjava.DelegatingDistributedLock.AcquireAction;
import com.coditory.sherlock.rxjava.DelegatingDistributedLock.ReleaseAction;
import io.reactivex.Single;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Duration;

import static com.coditory.sherlock.Preconditions.expectNonNull;

final class SherlockWithConnector implements Sherlock {
    private final Logger logger = LoggerFactory.getLogger(DistributedLockConnector.class);
    private final DistributedLockConnector connector;
    private final Duration defaultDuration;
    private final OwnerIdPolicy defaultOwnerIdPolicy;

    SherlockWithConnector(
        DistributedLockConnector connector,
        OwnerIdPolicy defaultOwnerIdPolicy,
        Duration defaultDuration
    ) {
        expectNonNull(connector, "connector");
        expectNonNull(defaultOwnerIdPolicy, "defaultOwnerIdPolicy");
        expectNonNull(defaultDuration, "defaultDuration");
        this.connector = connector;
        this.defaultOwnerIdPolicy = defaultOwnerIdPolicy;
        this.defaultDuration = defaultDuration;
    }

    @Override
    @NotNull
    public Single<InitializationResult> initialize() {
        logger.debug("Initializing sherlock infrastructure");
        return connector.initialize();
    }

    @Override
    @NotNull
    public DistributedLockBuilder<DistributedLock> createLock() {
        return createLockBuilder(connector::acquire, connector::release);
    }

    @Override
    @NotNull
    public DistributedLockBuilder<DistributedLock> createReentrantLock() {
        return createLockBuilder(connector::acquireOrProlong, connector::release);
    }

    @Override
    @NotNull
    public DistributedLockBuilder<DistributedLock> createOverridingLock() {
        return createLockBuilder(connector::forceAcquire, (id, __) -> connector.forceRelease(id));
    }

    @Override
    @NotNull
    public Single<ReleaseResult> forceReleaseAllLocks() {
        return connector.forceReleaseAll();
    }

    private DistributedLockBuilder<DistributedLock> createLockBuilder(
        AcquireAction acquireAction,
        ReleaseAction releaseAction
    ) {
        return new DistributedLockBuilder<>(createLock(acquireAction, releaseAction))
            .withLockDuration(defaultDuration)
            .withOwnerIdPolicy(defaultOwnerIdPolicy);
    }

    private LockCreator<DistributedLock> createLock(
        AcquireAction acquireAction,
        ReleaseAction releaseAction
    ) {
        return (lockId, duration, ownerId) ->
            createLockAndLog(acquireAction, releaseAction, lockId, ownerId, duration);
    }

    private DistributedLock createLockAndLog(
        AcquireAction acquireAction,
        ReleaseAction releaseAction,
        String lockId,
        String ownerId,
        Duration duration
    ) {
        DistributedLock lock = new DelegatingDistributedLock(
            acquireAction, releaseAction, lockId, ownerId, duration);
        logger.debug("Created lock: {}", lock);
        return lock;
    }
}

File: api/coroutines-connector/src/main/kotlin/com/coditory/sherlock/coroutines/SherlockWithConnector.kt
package com.coditory.sherlock.coroutines

import com.coditory.sherlock.DistributedLockBuilder
import com.coditory.sherlock.DistributedLockBuilder.LockCreator
import com.coditory.sherlock.OwnerIdPolicy
import com.coditory.sherlock.Preconditions.expectNonNull
import com.coditory.sherlock.coroutines.DelegatingDistributedLock.AcquireAction
import com.coditory.sherlock.coroutines.DelegatingDistributedLock.ReleaseAction
import org.slf4j.LoggerFactory
import java.time.Duration

class SherlockWithConnector(
    connector: SuspendingDistributedLockConnector,
    defaultOwnerIdPolicy: OwnerIdPolicy,
    defaultDuration: Duration,
) : Sherlock {
    private val logger = LoggerFactory.getLogger(SherlockWithConnector::class.java)
    private val connector: SuspendingDistributedLockConnector
    private val defaultDuration: Duration
    private val defaultOwnerIdPolicy: OwnerIdPolicy

    init {
        expectNonNull(connector, "connector")
        expectNonNull(defaultOwnerIdPolicy, "defaultOwnerIdPolicy")
        expectNonNull(defaultDuration, "defaultDuration")
        this.connector = connector
        this.defaultOwnerIdPolicy = defaultOwnerIdPolicy
        this.defaultDuration = defaultDuration
    }

    override suspend fun initialize() {
        logger.debug("Initializing sherlock infrastructure")
        connector.initialize()
    }

    override fun createLock(): DistributedLockBuilder<DistributedLock> {
        return createLockBuilder({ connector.acquire(it) }) { lockId: String, ownerId: String ->
            connector.release(lockId, ownerId)
        }
    }

    override fun createReentrantLock(): DistributedLockBuilder<DistributedLock> {
        return createLockBuilder({ connector.acquireOrProlong(it) }) { lockId: String, ownerId: String ->
            connector.release(lockId, ownerId)
        }
    }

    override fun createOverridingLock(): DistributedLockBuilder<DistributedLock> {
        return createLockBuilder({ connector.forceAcquire(it) }) { id: String, _: String ->
            connector.forceRelease(id)
        }
    }

    override suspend fun forceReleaseAllLocks(): Boolean {
        return connector.forceReleaseAll()
    }

    private fun createLockBuilder(
        acquireAction: AcquireAction,
        releaseAction: ReleaseAction,
    ): DistributedLockBuilder<DistributedLock> {
        return DistributedLockBuilder(createLock(acquireAction, releaseAction))
            .withLockDuration(defaultDuration)
            .withOwnerIdPolicy(defaultOwnerIdPolicy)
    }

    private fun createLock(
        acquireAction: AcquireAction,
        releaseAction: ReleaseAction,
    ): LockCreator<DistributedLock> {
        return LockCreator<DistributedLock> { lockId: String, duration: Duration?, ownerId: String ->
            createLockAndLog(acquireAction, releaseAction, lockId, ownerId, duration)
        }
    }

    private fun createLockAndLog(
        acquireAction: AcquireAction,
        releaseAction: ReleaseAction,
        lockId: String,
        ownerId: String,
        duration: Duration?,
    ): DistributedLock {
        val lock: DistributedLock =
            DelegatingDistributedLock(acquireAction, releaseAction, lockId, ownerId, duration)
        logger.debug("Created lock: {}", lock)
        return lock
    }
}
